---
title: "Analysis of particle size and the gut microbiome"
output: html_notebook
---

# Purpose:
Analyze fecal particle size (FPS) data to assess relationship with gut microbiome.

# Load libraries
```{r}
# Plotting
library(tidyverse)
library(RColorBrewer)
library(ggtext)
library(ggupset)
library(patchwork)
library(cowplot)

# Stats
library(phyloseq)
library(lme4)
library(lmerTest)
library(vegan)
library(ALDEx2)
library(ade4)
library(factoextra)
library(cluster)

```

## Additional setup
```{r}
# Set seed for reproducible analysis
set.seed(123)

# Set default ggplot theme
theme_set(theme_bw() +
            theme(axis.text = element_text(color = "black"),
                  axis.ticks = element_line(color = "black"),
                  plot.title = element_text(hjust = 0.5)))

## Favorite palette
paired_palette <- brewer.pal(n = 12, "Paired")

# Define functions
## Omit reads with obvious aggregation (particles greater than filter size detected)
omit_aggregates <- function(df) {
  df[rowSums(df[,which(colnames(df) == "X2100"):which(colnames(df) == "X3500")]) == 0,] %>%
    return()
}

## Summarize D10-D50-D90 per sample
### For each sample, we have multiple measurements, so we want to average these
summarize_fps_D <- function(df) {
  df %>%
    dplyr::select(participant:Dx..90.) %>%
    gather(key=D, value=val, (ncol(.)-2):ncol(.)) %>%
    mutate(D = case_when(D == "Dx..10." ~ "10th percentile",
                         D == "Dx..50." ~ "50th percentile (median)",
                         D == "Dx..90." ~ "90th percentile")) %>%
    group_by(participant, week, day_merge, D) %>%
    dplyr::summarize(mean=mean(val), se=sd(val)/sqrt(length(val)))
}

## Summarize all particle bins per sample
summarize_fps_all_bins <- function(df) {
  df %>%
    gather(key=size, value=volume, (ncol(.)-100):ncol(.)) %>%
    mutate(size = as.numeric(substr(size, 2, nchar(size)))) %>%
    group_by(participant, week, day_merge, size) %>%
    dplyr::summarize(mean=mean(volume)) %>%
    ungroup() %>%
    as.data.frame()
}

```

# Load data
```{r}
# Read in data
## Cohort 1 - Cognition study
cohort1 <- read.csv("data/onr_mastersizer_all.csv") %>%
  omit_aggregates() %>%
  dplyr::rename(day_merge = day) %>%
  mutate(week = "Baseline") %>%
  relocate(week, .after = participant) # make formatting consistent with cohort2

## Cohort 2 - Chewing study
cohort2 <- read.csv("data/epsom_mastersizer_with_metadata.csv") %>%
  omit_aggregates()

# Summarize D10-D50-D90 per sample
cohort1_D <- summarize_fps_D(cohort1)
cohort2_D <- summarize_fps_D(cohort2)

# Summarize all particle bins per sample
cohort1_all_bins <- summarize_fps_all_bins(cohort1)
cohort2_all_bins <- summarize_fps_all_bins(cohort2)

```

# Analyze and visualize FPS
## Plot baseline data for D10-D50-D90
```{r}
# Merge dataframes
D_all <- rbind(mutate(cohort1_D, cohort = "Cohort 1"),
               mutate(cohort2_D, cohort = "Cohort 2",
                      day_merge = as.character(day_merge))) %>%
  filter(week == "Baseline")

# Arrange by median value per participant for organized plot
baseline_order <- D_all %>%
  filter(D == "50th percentile (median)") %>%
  group_by(participant) %>%
  dplyr::summarize(baseline_mean = mean(mean)) %>%
  arrange(baseline_mean) %>%
  pull(participant)

# Statistical tests
## ANOVA of median by participant
D_all %>%
  filter(D == "50th percentile (median)") %>%
  aov(mean ~ participant, data = .) %>%
  summary()
# participant  75 208005  2773.4   8.983 <2e-16 ***

## Average median FPS across all baseline samples
D_all %>%
  filter(D == "50th percentile (median)") %>%
  pull(mean) %>%
  mean() # 43.1 um

# Plot baselines
D_baseline_plot <- D_all %>%
  mutate(participant = factor(participant, levels = baseline_order)) %>%
  ggplot(aes(x = mean, y = participant, group = participant, color = cohort)) +
  geom_point() +
  geom_line(linetype="dotted") +
  geom_errorbar(aes(xmin=mean-se, xmax=mean+se), width=0.5) +
  labs(y="Participant", x= "Particle size (μm)", color = "Cohort") +
  scale_color_manual(values = c(paired_palette[2], paired_palette[6])) +
  theme(axis.text.y=element_blank(),
        axis.ticks.y=element_blank()) +
  facet_wrap(~D, scales="free_x") +
  scale_x_log10() +
  geom_textbox(data = data.frame(x=15, y = 66, label = "ANOVA<br>*p* < 2 × 10<sup>-16</sup>",
                                 D = "50th percentile (median)"),
               aes(x, y, label = label), halign = 0.5, fill = NA, box.colour = NA,
               box.padding = grid::unit(rep(0, 4), "pt"), inherit.aes = F)
#ggsave("plots/baseline_FPS_both_cohorts.png", D_baseline_plot, height = 6, width = 9)

D_baseline_plot

```

## Plot all particle size bins
```{r}
# Merge data frames
all_bins_all <- rbind(mutate(cohort1_all_bins, cohort = "Cohort 1"),
               mutate(cohort2_all_bins, cohort = "Cohort 2",
                      day_merge = as.character(day_merge))) %>%
  filter(week == "Baseline") %>%
  mutate(sample = paste0(participant, "_", day_merge))

# Calculate mean and se
all_bins_summary <- all_bins_all %>%
  group_by(size) %>%
  summarize(mean2 = mean(mean), se = sd(mean)/sqrt(length(mean)), sd = sd(mean))

# Plot all curves
all_curves_plot <- ggplot(all_bins_all, aes(x = size, y = mean, group = sample)) +
  geom_line(alpha = 0.175) +
  geom_ribbon(data = all_bins_summary, aes(x = size, ymin=mean2-sd, ymax=mean2+sd),
              color = NA, fill = "red", alpha = 0.25, inherit.aes = F) +
  geom_line(data = all_bins_summary, aes(x = size, y = mean2),
            color = "red3", size = 1.25, inherit.aes = F) +
  scale_x_log10(limits=c(0.276, 2100)) +
  labs(x="Particle size (μm)", y="Particles by volume (%)") +
  geom_hline(color = "black", yintercept = 0, linetype = "dashed")

all_curves_plot

# Do PCA ordination
all_bins_spread <- all_bins_all %>%
  spread(key = size, value = mean)

all_bins_spread_data <- all_bins_spread %>%
  dplyr::select(`0.01`:`3500`) %>%
  select_if(colSums(.) > 0)

fps_pca <- prcomp(all_bins_spread_data, center = TRUE, scale. = TRUE)

# slow step
# adonis2(scale(all_bins_spread_data) ~ cohort + participant, data = all_bins_spread,
#       permutations=9999, method = "eu")
# participant: R2 = 0.74 p < 0.0001
# cohort: 

## Add median to include size in PCA
D_all_for_merge <- D_all %>%
  filter(D == "50th percentile (median)") %>%
  mutate(sample = paste0(participant, "_", day_merge)) %>%
  dplyr::rename(median_FPS = mean)
  
## Need a version of this dataframe with metadata
all_bins_spread <- all_bins_spread %>%
  mutate(sample = paste0(participant, "_", day_merge)) %>%
  left_join(D_all_for_merge)

## Plot PCA
fps_pca_plot <- autoplot(fps_pca, data = all_bins_spread, colour = "participant", size = "median_FPS", alpha = 0.5) +
  geom_line(aes(group = participant, color = participant)) +
  theme(legend.position = "none") +
  geom_textbox(data = data.frame(x=0.1, y = -0.2, label = "PERMANOVA<br>*R*<sup>2</sup> = 0.74<br>*p* < 0.0001"),
               aes(x, y, label = label), halign = 0.5, fill = NA, box.colour = NA,
               box.padding = grid::unit(rep(0, 4), "pt"), inherit.aes = F)
#ggsave("plots/baseline_FPS_PCA.png", fps_pca_plot, height = 4, width = 5)


```

## K-means clustering
```{r}
# How many clusters should we have?
## Sum of squares plot
sos_plot <- fviz_nbclust(all_bins_spread_data, kmeans, method = "wss") +
  theme_bw() + # elbow at k=4-5
  labs(y = "Total within sum of square") +
  theme(plot.title = element_blank())

## Gap stat plot (~10 seconds)
gap_stat <- clusGap(all_bins_spread_data, FUN = kmeans, nstart = 25, K.max = 10, B = 50) %>%
   fviz_gap_stat() + # slower method; max at k=5
  theme_bw() + theme(plot.title = element_blank())

sos_plot + gap_stat

# Perform k-means clustering with k = 5 clusters
## Determine clusters
km <- kmeans(all_bins_spread_data, centers = 5, nstart = 25)

## Plot PCA by cluster
k_clust <- fviz_cluster(km, data = all_bins_spread_data, labelsize = NA, main = F,
                        shape = 19, show.clust.cent = F,
                        pointsize = log2(all_bins_spread$median_FPS) / 2, alpha = 0.5) +
  theme_bw() +
  labs(color = "Cluster", fill = "Cluster", shape = "Cluster")

k_clust

## Keep data about which sample maps to which cluster
kdata <- cbind(all_bins_spread[,c(1:5, 108)], cluster = km$cluster)

## Summarize within cluster (mean + se)
### Median
kdata_summary <- kdata %>%
  group_by(cluster) %>%
  summarize(mean = mean(median_FPS), se=sd(median_FPS)/sqrt(length(median_FPS)))

### Curves
kcurves_summary <- all_bins_all %>%
  left_join(kdata[,c("sample", "cluster")]) %>%
  mutate(cluster = as.character(cluster)) %>%
  group_by(cluster, size) %>%
  summarize(mean2 = mean(mean), se=sd(mean)/sqrt(length(mean)))

# Plot curves
k_curves_summary_plot <- ggplot(kcurves_summary, aes(x = size, y = mean2,
                                                     group = cluster, color = cluster)) +
  geom_ribbon(data = kcurves_summary, aes(x = size, ymin=mean2-se, ymax=mean2+se,
                                          fill = cluster),
              color = NA, alpha = 0.35, inherit.aes = F) +
  geom_line() +
  labs(x="Particle size (μm)", y="Particles by volume (%)",
       color = "Cluster", fill = "Cluster") +
  geom_hline(color = "black", yintercept = 0, linetype = "dashed") +
  scale_x_log10(limits=c(0.276, 2100))

k_curves_summary_plot

# Analyze clusters by median FPS
kdata %>%
  aov(median_FPS ~ factor(cluster) + participant, data = .) %>%
  summary()
# cluster p < 2e-16, participant p = 0.00034

## Follow up with TukeyHSD test (~which~ clusters are different?)
kdata %>%
  aov(median_FPS ~ factor(cluster) + participant, data = .) %>%
  TukeyHSD(which = "factor(cluster)")
# Tukey: all diff except 4-5

# Plot median_FPS
k_median_plot <- ggplot(kdata, aes(x = cluster, y = median_FPS)) +
  geom_point(position=position_jitter(0.05), alpha=0.25) +
  geom_errorbar(data = kdata_summary, aes(x=cluster, ymin=mean, ymax=mean),
                color = "red", width=0.5, inherit.aes = F) +
  geom_errorbar(data = kdata_summary, aes(x=cluster, ymin=mean-se, ymax=mean+se),
                width=0.25, inherit.aes = F) + 
  labs(x = "Cluster", y = "Median FPS (μm)") +
  annotate(geom = "text", x = 1:5, y = c(20, 250, 120, 90, 60),
           label = c("a", "d", "c", "b", "b")) +
  scale_y_log10()

k_median_plot

# Upset plot (do samples from a given participant land in multiple clusters)
k_upset <- kdata %>%
  dplyr::select(participant, cluster) %>%
  distinct() %>%
  group_by(participant) %>% 
  summarize(clusters = list(cluster)) %>% 
  ggplot(aes(x=clusters)) +
  geom_bar() +
  scale_x_upset(n_intersections = 20) +
  labs(x = "Clusters", y = "Participants (#)")
# 39 participants in only one cluster

k_upset + k_curves_summary_plot

```


# Relating baseline FPS to 16S
## 16S - diversity and overall NMDS
```{r}
ps <- readRDS("data/phyloseq_EPSOM_and_ONR.rds")

ps
# Start with 3673 taxa, 313 samples

ps_filt <- prune_samples(sample_sums(ps) >= 5000, ps) %>%
  filter_taxa(function(x) sum(x > 3) > 0.1*length(x), TRUE)

ps_filt
# Reduces to 287 taxa, 313 samples

# Extract sample data from phyloseq object for easy data wrangling later
samdf <- sample_data(ps) %>%
  as.data.frame()

# Examine diversity
# Using unfiltered data here
diversity <- ps %>%
  estimate_richness(split = TRUE, measures=c("Shannon", "Observed")) %>%
  cbind(samdf[rownames(.),], .) %>%
  gather(key = measure, value = value, c("Shannon", "Observed")) %>%
  mutate(sample = paste0(participant, "_", day)) %>%
  relocate(sample) %>%
  inner_join(D_all_for_merge)
# SM56_F1 not present in 16S - seems there was a problem with that well

## lm
diversity %>%
  filter(measure == "Observed") %>%
  lmer(value ~ median_FPS + (1 | participant), data = .) %>%
  summary()
# Observed: p = 4.75e-09

diversity %>%
  filter(measure == "Shannon") %>%
  lmer(value ~ median_FPS + (1 | participant), data = .) %>%
  summary()
# Shannon: p = 5.37e-07

# Get total read depth
## Checking if read depth could be driving the observed diversity-FPS correlation
## (i.e. if there are more bacterial cells, we get more reads --> more diversity)
read_depth <- ps %>%
  psmelt() %>%
  mutate(sample = paste0(participant, "_", day)) %>%
  group_by(sample) %>%
  dplyr::summarize(reads = sum(Abundance))

diversity <- diversity %>%
  left_join(read_depth)

# Get diversity and FPS means within participant
diversity_summary <- diversity %>%
  group_by(participant, cohort, measure) %>%
  summarize(value = mean(value), median_FPS = mean(median_FPS), reads = mean(reads))

diversity_summary %>%
  filter(measure == "Observed") %>%
  lm(value ~ median_FPS, data = .) %>%
  summary()
# Observed R2 = 0.3712

diversity_summary %>%
  filter(measure == "Shannon") %>%
  lm(value ~ median_FPS, data = .) %>%
  summary()
# Shannon R2 = 0.3359

## Spearman
cor.test(diversity_summary[diversity_summary$measure == "Observed",]$value,
         diversity_summary[diversity_summary$measure == "Observed",]$median_FPS,
         method = "spearman")
# Observed: p = 5.617e-10, rho = -0.6381527

cor.test(diversity_summary[diversity_summary$measure == "Shannon",]$value,
         diversity_summary[diversity_summary$measure == "Shannon",]$median_FPS,
         method = "spearman")
# Shannon: p = 3.04e-08, rho = -0.5940123 


## Plot diversity and correlation
diversity_summary_plot <- ggplot(diversity_summary,
                                 aes(x = median_FPS, y = value, color = cohort)) +
  geom_point() +
  scale_color_brewer(palette = "Set1") +
  labs(x = "Median FPS (μm)", y = "Alpha diversity", color = "Cohort") +
  geom_textbox(data = data.frame(x=c(62, 62),
                                 y = c(220, 4.2),
                                 label = c("Spearman's &rho; = -0.64<br>*p* = 5.6 x 10<sup>-10</sup>",
                                           "Spearman's &rho; = -0.59<br>*p* = 3.0 x 10<sup>-8</sup>"),
                                 measure = c("Observed", "Shannon")),
               aes(x, y, label = label), halign = 0.5, fill = NA, box.colour = NA,
               box.padding = grid::unit(rep(0, 4), "pt"), inherit.aes = F) +
  facet_wrap(~measure, scales = "free_y") +
  scale_x_log10()

diversity_summary_plot


# Diversity on PCA
diversity_for_merge <- diversity %>%
  dplyr::select(sample, measure, value) %>%
  spread(key = measure, value = value)
  
all_bins_spread_div <- all_bins_spread %>%
  inner_join(diversity_for_merge)

all_bins_spread_data_div <- all_bins_spread_div %>%
  dplyr::select(`0.01`:`3500`) %>%
  select_if(colSums(.) > 0)
 
fps_pca_div <- prcomp(all_bins_spread_data_div, center = TRUE, scale. = TRUE)
 
# adonis2(scale(all_bins_spread_data_div) ~ Observed + participant,
#         data = all_bins_spread_div,
#         permutations=9999, method = "eu")
# Shannon       1   1626.2 0.14255 58.7333  1e-04 ***
# participant  75   6791.4 0.59532  3.2704  1e-04 ***

# Observed      1   1968.6 0.17256 71.654  1e-04 ***
# participant  75   6472.2 0.56734  3.141  1e-04 ***
 
diversity_pca_obs <- autoplot(fps_pca_div, data = all_bins_spread_div,
                         fill = "Observed", size = "median_FPS",
                         shape = 21, color = "black") +
  scale_fill_distiller(palette = "RdYlBu", direction = -1) +
  labs(fill = "Observed ASVs", size = "Median FPS (μm)") +
  geom_textbox(data = data.frame(x=0.05, y = 0.18, label = "PERMANOVA<br>observed *R*<sup>2</sup> = 0.17, *p* < 0.0001<br>participant *R*<sup>2</sup> = 0.57, *p* < 0.0001"),
               aes(x, y, label = label), halign = 0.5, fill = NA, box.colour = NA,
               box.padding = grid::unit(rep(0, 4), "pt"), 
               width = unit(3, "inch"), size = 3.5, inherit.aes = F)

diversity_pca_sha <- autoplot(fps_pca_div, data = all_bins_spread_div,
                         fill = "Shannon", size = "median_FPS",
                         shape = 21, color = "black") +
  scale_fill_distiller(palette = "RdYlBu", direction = -1) +
  labs(fill = "Shannon diversity", size = "Median FPS (μm)") +
  geom_textbox(data = data.frame(x=0.05, y = 0.17, label = "PERMANOVA<br>Shannon *R*<sup>2</sup> = 0.14, *p* < 0.0001<br>participant *R*<sup>2</sup> = 0.60, *p* < 0.0001"),
               aes(x, y, label = label), halign = 0.5, fill = NA, box.colour = NA,
               box.padding = grid::unit(rep(0, 4), "pt"),
               width = unit(3, "inch"), size = 3.5, inherit.aes = F)


diversity_pca_obs + diversity_pca_sha

# Microbiome NMDS with FPS
ps_baseline <- ps %>%
  subset_samples(week == "Baseline")

samdf_baseline <- sample_data(ps_baseline) %>%
  as.matrix() %>%
  as.data.frame() %>%
  mutate(sample = paste0(participant, "_", day)) %>%
  left_join(dplyr::select(D_all_for_merge, sample, median_FPS))

rownames(samdf_baseline) <- sample_names(ps_baseline)

ps_baseline@sam_data <- samdf_baseline %>%
  sample_data()

ps_baseline <- ps_baseline %>%
  subset_samples(!is.na(median_FPS))

# Extract points for greater freedom of plotting
ord.nmds.bray <- ordinate(ps_baseline, method="NMDS", distance="bray")

nmds <- ord.nmds.bray$points %>% as.data.frame() %>%
  cbind(as.matrix(as.data.frame(ps_baseline@sam_data)), .) %>%
  mutate(median_FPS = as.numeric(median_FPS)) %>%
  left_join(kdata[,c("sample", "cluster")]) %>%
  mutate(cluster = factor(cluster))


bray <- distance(ps_baseline, method = "bray") %>% as.matrix()
# slow
# adonis2(formula = bray ~ cluster + cohort + participant, 
#          data = samdf_baseline[rownames(bray),] %>%
#           left_join(kdata[,c("sample", "cluster")]) %>%
#           mutate(cluster = factor(cluster)),
#          permutations=9999, method = "bray")
#cluster       4    2.818 0.05446  9.0699  1e-04 ***
#cohort        1    3.757 0.07259 48.3577  1e-04 ***
#participant  74   37.023 0.71535  6.4403  1e-04 ***


nmds_cluster_plot <- ggplot(nmds, aes(x=MDS1, y=MDS2, size = median_FPS, color = cluster)) +
  geom_point(alpha = 0.5) +
  labs(x="NMDS1", y = "NMDS2", size="Median\nFPS (μm)", color = "Cluster") +
  geom_textbox(data = data.frame(x=0.835, y = -1, label = "PERMANOVA<br>cluster *R*<sup>2</sup> = 0.054, *p* < 0.0001<br>cohort *R*<sup>2</sup> = 0.073, *p* < 0.0001<br>participant *R*<sup>2</sup> = 0.72, *p* < 0.0001"),
               aes(x, y, label = label), halign = 0.5, fill = NA, box.colour = NA,
               box.padding = grid::unit(rep(0, 4), "pt"),
               width = unit(3, "inch"), inherit.aes = F)

nmds_cluster_plot


```

## 16S - specific taxa
```{r}
ps_baseline_filt <- ps_baseline %>%
  prune_samples(sample_sums(.) >= 5000, .) %>%
  filter_taxa(function(x) sum(x > 3) > 0.1*length(x), TRUE)

# ALDEx2 KW test by cluster
conds <- samdf_baseline[rownames(as.data.frame(as.matrix(ps_baseline_filt@sam_data))),] %>%
  left_join(kdata[,c("sample", "cluster")]) %>%
  mutate(cluster = factor(cluster)) %>%
  pull(cluster)

#kw.test <- aldex.clr(otu_flip, conds, mc.samples=128, denom="all") %>%
#  aldex.kw()
#write.csv(kw.test, "plots/aldex_kw_cluster.csv")

# 29 / 298 KW.eBH p < 0.05

kw.test <- read.csv("plots/aldex_kw_cluster.csv", row.names = 1)


melty <- ps_baseline_filt %>%
  microbiome::transform(transform = "compositional") %>%
  psmelt() %>%
  left_join(kdata[,c("sample", "cluster")]) %>%
  mutate(cluster = factor(cluster)) %>%
  mutate(binom = case_when(is.na(Genus) ~ paste0(Family, " family (", OTU, ")"),
                           !is.na(Genus) & is.na(Species) ~
                             paste0("*", Genus, " sp.* (", OTU, ")"),
                           !is.na(Species) ~
                             paste0("*", Genus, " ", Species, "* (", OTU, ")")))

select_taxa <- c("ASV 121", "ASV 30", "ASV 42") # top three by kw.eBH

select_taxa_order <- c("*UCG-002 sp.* (ASV 121)",
                       "*Anaerostipes hadrus* (ASV 30)",
                       "*[Ruminococcus] gnavus group sp.* (ASV 42)")

melty_summary <- melty %>%
  filter(OTU %in% select_taxa) %>%
  group_by(binom, cluster) %>%
  summarize(mean = mean(Abundance), se=sd(Abundance)/sqrt(length(Abundance))) %>%
  mutate(binom = factor(binom, levels = select_taxa_order))

# Follow up with Tukey test on CLR
ps_baseline_filt %>%
  microbiome::transform(transform = "clr") %>%
  psmelt() %>%
  filter(OTU == "ASV 30") %>%
  left_join(kdata[,c("sample", "cluster")]) %>%
  mutate(cluster = factor(cluster)) %>%
  aov(Abundance ~ cluster + participant, data = .) %>%
  TukeyHSD(which = "cluster")
# ASV 121: all diff except 2-4 and 3-5 --> a, b, c, b, c
# ASV 30: all diff except 1-2, 1-4, 2-4 --> a, a, b, a, c
# ASV 42: all diff except 1-2, 1-4, 2-4, 3-5 --> a, a, b, a, b

# Plot top three taxa
tax_plot <- melty %>%
 filter(OTU %in% select_taxa) %>%
  mutate(binom = factor(binom, levels = select_taxa_order)) %>%
 ggplot(aes(x = cluster, y = Abundance, color = cohort)) +
 geom_point(position = position_jitter(height = 0, width = 0.05), alpha = 0.5) +
 geom_errorbar(data = melty_summary, aes(x=cluster, ymin=mean, ymax=mean),
               color = "red", width=0.5, inherit.aes = F) +
 geom_errorbar(data = melty_summary, aes(x=cluster, ymin=mean-se, ymax=mean+se),
               width=0.25, inherit.aes = F) + 
 labs(x = "Cluster", y = "Relative abundance", color = "Cohort") +
 scale_color_brewer(palette = "Set1") +
 facet_wrap(~binom, scales = "free", ncol = 1) +
 theme(strip.text = element_markdown())


```
